### PR TITLE
Added Jenkins And Docker

### PR DESCRIPTION
**Pull request by tavisca-pkadam**

### CODE REVIEW

**Comment by tavisca-pkadam** [comment]
@tavisca-nkumar

**Comment by tavisca-pkadam** [comment]
The Web-Api doesnt contains test cases so the jenkins file only contains build stage 
`https://github.com/tavisca-pkadam/CartingApp/blob/dev/Jenkinsfile`
 link contains build and test

**Comment by tavisca-pkadam** [comment]
Updated Jenkins File @tavisca-nkumar

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            }
+            steps {
+                bat '''dotnet build %SOLUTION_NAME%'''
+            }
+        }
+
+        
+        stage('sonar') {
+             when {
+                expression { return params.SONAR_ANALYSIS}
```
Comment (ghost) [inline Jenkinsfile]: why need a return here?

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                        dotnet ${SONAR_MS_TOOL}  begin /k:"%SONAR_PROJECT_NAME%" /d:sonar.host.url=${SONAR_URL}  /d:sonar.login="${SONAR_TOKEN}"
+                        dotnet  build
+                        dotnet ${SONAR_MS_TOOL} end  /d:sonar.login="${SONAR_TOKEN}"
+                    """
+            }     
+        }
+
+        stage('publish') {
+             when {
+                expression { return params.PUBLISH || params.DEPLOY }
```
Comment (ghost) [inline Jenkinsfile]: why deploy condition is needed?

File: Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            when {
+                  expression {return params.DOCKER_HUB_PUBLISH}
+            }
+            
+            steps {
+             script{
+                    docker.withRegistry('','docker_creds')
+                    {
+                        
+                         bat ''' docker tag %DOCKER_IMAGE_NAME%:latest subtleparesh/%DOCKER_IMAGE_NAME%
```
Comment (ghost) [inline Jenkinsfile]: docker image versioning is missing and docker registry name is hardcoded.

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                    docker.withRegistry('','docker_creds')
+                    {
+                        
+                         bat ''' docker tag %DOCKER_IMAGE_NAME%:latest subtleparesh/%DOCKER_IMAGE_NAME% 
+                        docker push subtleparesh/%DOCKER_IMAGE_NAME%:latest'''
+                    }
+                }
+            }
+        }
+         stage('run docker image'){
```
Comment (ghost) [inline Jenkinsfile]: stage name is not appropriate

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+             when {
+                  expression {return params.DOCKER_RUN}
```
Comment (ghost) [inline Jenkinsfile]: need of docker run condition?

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            }
+            steps{
+                echo 'run the image'
+                bat 'docker run -p %PORT_NO%:80 -e SOLUTION_DLL=%SOLUTION_DLL%  %DOCKER_IMAGE_NAME%'
```
Comment (ghost) [inline Jenkinsfile]: container port is should be parameterized and docker image should be pulled from docker repo.

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,9 @@
+FROM mcr.microsoft.com/dotnet/core/aspnet:2.2
+
+ENV SOLUTION_DLL="WebApi.dll"
+
+WORKDIR /home_dir
+
+COPY  ./artifacts/. .
```
Comment (ghost) [inline Dockerfile:7]: what if my directory name is not artifacts?

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+
+ENTRYPOINT dotnet ${SOLUTION_DLL}
```
Comment (ghost) [inline Dockerfile:9]: good use of env

**Review by ghost — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 7 above)
Comment (tavisca-pkadam) [reply Dockerfile:7]: all artifacts are created in the "artifacts" folder for all solutions
 -o ../artifacts

**Review by ghost — APPROVED** [review]

**Pull request merged**